### ISSUE TITLE
FormCommit context menu order changed

### ISSUE THREAD

**Issue by RussKie** [gitextensions/gitextensions#8106]
## Current behaviour

I'm finding the new order extremely annoying - now to reset a file or a change I need to move the mouse. I can see a diff in the `FormCommit`, I don't need to open an external diff tool for that.

The new behaviour (v3.4):
![image](https://user-images.githubusercontent.com/4403806/81927183-91f68f80-9626-11ea-9ba2-8149d635ded6.png)

The old behaviour (v3.3.1 and below):
![image](https://user-images.githubusercontent.com/4403806/81927248-ae92c780-9626-11ea-9832-fbb2562dfd32.png)

I'd like to have the order changed to a more logical grouping:
![image](https://user-images.githubusercontent.com/4403806/81927943-d0d91500-9627-11ea-8b9c-a10519b70603.png)

**Comment by gerhardol** [comment]
The menus were aligned in #7961 (and #8065) for most menus
If changed in commit, it should be changed in RevDiff too

I definitely use difftool more than reset, the default diff is limited and would like to have difftool as the default double click in RevDiff, why I believe on top is the best position.
(I was initially annoyed by GE having stage/unstage in FormCommit by default double click in addition to RevDiff behavior that I am still annoyed by.)

If moved, difftool should be at the top of the second group

**Comment by mstv** [comment]
It happened to me too: `Reset file or directory changes` was not easy to find today.
Though I think I could get used to the new position. The icon helps and it has a shortcut -- `R` works also in the RevDiff in Gerhard's and my build.  

Might be also caused by: Why "... or directory ..."? Git does not handle directories (unfortunately).
If there is submodule-specific handling, it should rather be named `Reset file or submodule changes`.
Else it should be just `Reset file(s)` as in the RevDiff. 

> If changed in commit, it should be changed in RevDiff too

👍 But this makes me strongly vote against the move.
It would be a nuisance to open the often used submenu if `Open with difftool` was not the first item:

![grafik](https://user-images.githubusercontent.com/36601201/82094404-46fb7b80-96fd-11ea-8d74-3fe2ee253dda.png)

> If moved, difftool should be at the top of the second group

Yes, this would be the most logical position IMO if any item was moved.
But I don't want `Reset` and `Cherry pick` below the `Open`/`Edit` group.
Instead I propose to add a separator below `Open with difftool`.

> I ... would like to have difftool as the default double click in RevDiff

I agree, I need the history much less (but open it by mistake quite often).

**Comment by mstv** [comment]
What do you think about the following compromise?

![grafik](https://user-images.githubusercontent.com/36601201/82160971-5a434e00-9899-11ea-874c-47cc73f1838e.png)

**Comment by gerhardol** [comment]
> What do you think about the following compromise?

The text should be changed regardless, align with RevDiff etc
You can reset submodules with some limitations, but you most likely what to reset what is _within_ the submodule, and for worktree you have that as a separate option

I prefer to not change the order at all, if changed I propose to move it to top of next group for all forms

**Comment by RussKie** [comment]
![image](https://user-images.githubusercontent.com/4403806/82730404-aaf7f280-9d42-11ea-8eef-351ccf55b094.png)

I've compared the context menus I could think of (is there more?).
Indeed we have consistency, but I am unconvinced this consistency is sensible for `FormCommit` as it has different functions to other presentations.

In the File Tree and Diff tabs a user would typically deal with existing commits (i.e. snapshots), and "compare" function is likely to be one of the most used (i.e. compare to the working directory).
In the Commit form the user is dealing with the working directory changes, and thus staging and reseting is likely be the primary function, then compare (which is rendered in the diff panel of the form).

**Comment by RussKie** [comment]
Btw, we have "View history" in all menus except in Diff's version, which uses "File history". It also uses "Save selected as..." instead of "Save as...".
Something to fix in the master.

**Comment by gerhardol** [comment]
> I've compared the context menus I could think of (is there more?).

See #7961 (and #8065)
For instance submodules extra menus are missing.
I tried to cleanup this in #7961 but did not get all the way, similar with #8136.

> Indeed we have consistency, but I am unconvinced this consistency is sensible for FormCommit as it has different functions to other presentations.

Before #7961, the order was not aligned for the different context menus and I often had to scan the menu to find a specific item (non motivated differences).
There are sure reasons to have different options in the context menus but different order should be motivated.
Sure, we could switch Diff and Reset order, but I prefer them this way. #8137 only changes order in FormCommit.

I though have an argument that contradict myself: I want the doubleclick action to be dependent of the form:
 * Commit stage/unstage
 * Revdiff: difftool for file, open for submodules
 * FileTree: History for files, open for submodules
Ideally this action would be on top of the menu, but that would disrupt the order quite a lot. Reset would still not be on top.

> Btw, we have "View history" in all menus except in Diff's version, which uses "File history". It also uses "Save selected as..." instead of "Save as...".
Something to fix in the master.

This could be fixed in 3.4, such a cleanup is not essential to the translation. But I rather get 3.4 out now...
RevDiff is a little special, as more than one commit can be selected and only the last selected appear in the context menu. For reset, there is a submenu, so no problem. For the other forms, _Reset file(s)_ was suggested. There is no multi selection in RevFile, do we have to use plurals there or is it OK to ignore this?

**Comment by RussKie** [comment]
> This could be fixed in 3.4, such a cleanup is not essential to the translation. But I rather get 3.4 out now...
> There is no multi selection in RevFile, do we have to use plurals there or is it OK to ignore this?

Yes, I'm just calling out the differences that I've noticed having all of these on the screen.
These can wait for the next release.

**Comment by mstv** [comment]
Regarding #8137: I am undecided regarding the (non-)consistent order of context menu items.
Consistency of crowded context menus has a high value on the one hand.  On the other hand, different use cases can have other most used items, which should be easy to reach. Though even we do have different opinions about which should be the first item. (I do not have a strong opinion about FormCommit, since my free hand hits `F3` quicker than the other clicks the context menu.)

Because `Open with difftool` is not related to `Reset file` (and `Cherry pick`), I still propose to insert a separator instead of changing the order.

The default context menu item (double click), does not necessarily need to be the first context menu item IMO.

**Comment by RussKie** [comment]
> Because `Open with difftool` is not related to `Reset file` (and `Cherry pick`), I still propose to insert a separator instead of changing the order.

Personally I would put it with the rest of "Open" entries.
Something like:
```
Open with difftool
Open
Open with...
...(other Open)..
Edit
Save
```

I wouldn't consider this a release-blocking issue, but happy to have it actioned if you feel strongly about it.

**Comment by gerhardol** [comment]
Closed by #8175
